File: tech-studio-projects/financial-industry/crypto-trading-simulation/java/src/network/p2p/P2PClient.cpp
#include "P2PClient.h"
#include <cstring>
#include <unistd.h>
#include <iostream>

P2PClient::P2PClient() : sock(-1) {
}

P2PClient::~P2PClient() {
    if (sock != -1) {
        close(sock);
    }
}

bool P2PClient::connect(const std::string& ip, int port) {
    sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) {
        std::cerr << "Socket creation failed" << std::endl;
        return false;
    }
    
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(port);
    inet_pton(AF_INET, ip.c_str(), &serverAddr.sin_addr);
    
    if (::connect(sock, (struct sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
        std::cerr << "Connection failed" << std::endl;
        return false;
    }
    
    return true;
}

bool P2PClient::sendMessage(const std::string& message) {
    return send(sock, message.c_str(), message.size(), 0) != -1;
}

std::string P2PClient::receiveMessage() {
    char buffer[1024] = {0};
    int bytesRead = recv(sock, buffer, sizeof(buffer), 0);
    return bytesRead > 0 ? std::string(buffer, bytesRead) : "";
}

File: tech-studio-projects/financial-industry/cryptocurrency-trading-simulation/java/src/network/p2p/P2PClient.java
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;

public class P2PClient {

    private Socket socket;

    // Constructor
    public P2PClient() {
        socket = null;
    }

    // Method to connect to a server
    public boolean connect(String ip, int port) {
        try {
            socket = new Socket(ip, port);
            return true;
        } catch (IOException e) {
            System.err.println("Connection failed: " + e.getMessage());
            return false;
        }
    }

    // Method to send a message to the server
    public boolean sendMessage(String message) {
        if (socket == null || socket.isClosed()) {
            System.err.println("Socket is not initialized or is closed.");
            return false;
        }

        try {
            OutputStream outputStream = socket.getOutputStream();
            outputStream.write(message.getBytes());
            outputStream.flush();
            return true;
        } catch (IOException e) {
            System.err.println("Failed to send message: " + e.getMessage());
            return false;
        }
    }

    // Method to receive a message from the server
    public String receiveMessage() {
        if (socket == null || socket.isClosed()) {
            System.err.println("Socket is not initialized or is closed.");
            return "";
        }

        try {
            InputStream inputStream = socket.getInputStream();
            byte[] buffer = new byte[1024];
            int bytesRead = inputStream.read(buffer);
            return bytesRead > 0 ? new String(buffer, 0, bytesRead) : "";
        } catch (IOException e) {
            System.err.println("Failed to receive message: " + e.getMessage());
            return "";
        }
    }

    // Cleanup method to close the socket
    public void close() {
        if (socket != null && !socket.isClosed()) {
            try {
                socket.close();
            } catch (IOException e) {
                System.err.println("Failed to close socket: " + e.getMessage());
            }
        }
    }
}
